src/test/run-pass/panic-uninitialized-zeroed.rs

```diff
@@ -0,0 +1,31 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// This test checks that instantiating an uninhabited type via `mem::{uninitialized,zeroed}` results
+// in a runtime panic.
+
+#![feature(never_type)]
+
+use std::{mem, panic};
+
+struct Foo {
+    x: u8,
+    y: !,
+}
```

Comment:
Could you add a test for an empty enum?

Also, AFAIK `Foo` is NOT layout-uninhabited, because of https://github.com/rust-lang/rust/issues/49298?

Reply:
Correct, you need to change `x`'s type to `()`.

Reply:
Wait, no, check this out:
```rust
#![feature(never_type)]
pub struct Foo { x: u8, y: ! }
pub fn foo() -> Foo { loop {} }
```
That gets `noreturn` in the LLVM IR, with `u8` or `()` in the first field, but if you change the `!` to `()`, you lose the `noreturn`.

What happened in the first for #49298 is we *kept* the layout-uninhabited property *propagated*, but it wouldn't affect type sizes the same.

Which means we can use the check here to detect `noreturn` UB in all relevant cases!

Reply:
`nocall` would be more correct than `noreturn` though :P

Which "check here"?

Reply:
I mean the panicking intrinsics.